# **Final Project**

- Group: 7

- Member: 
    + 20127214: Nguyễn Trương Minh Khôi
    + 20127612: Đỗ Khánh Sang

---

## **Import**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import missingno as msno
import re
from datetime import datetime, timedelta

---

## **1.Collecting Data**

### **1.What subject is your data about?**

Main subject of my data is about reviews all kind of wine around the world such as: the number of points WineEnthusiast rated the wine, the country that the wine is from, the winery that made the wine, the cost for a bottle of the wine,…

### **2.What is the source of your data?**

My team take dataset from [Kaggle](https://www.kaggle.com/)

### **3.Do authors of this data allow you to use like this? You can check the data license**

- The authors totally agree me to use this data because this is a public dataset on kaggle and I can download it
- Data license: `CC BY-NC-SA 4.0`

### **4.How did authors collect data?**

The data was scraped from `WineEnthusiast` during the week of June 15th, 2017. The code for the scraper can be found [here](https://github.com/zackthoutt/wine-deep-learning) if you have any more specific questions about data collection that I didn't address.

## 2.Exploring Data 

### **1. Load data**

First, read data file "winemag-data-130k-v2.csv" from folder `Data` and save to DataFrame `wine_df`.

In [2]:
# load data
wine_df = pd.read_csv('Data/winemag-data-130k-v2.csv', index_col = 0)
wine_df.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

### **2. How many rows and how many columns?**

Next, get the number of rows and columns of the DataFrame `wine_df` and store it in `num_rows` and `num_cols` variables.

In [3]:
num_rows = len(wine_df)
num_columns = len(wine_df.columns)
print('Number of rows:', num_rows)
print('Number of columns', num_columns)

Number of rows: 129971
Number of columns 13


### **3. What is the meaning of each row?**

**Each row provides diferent information such as price, country, winery,... about different wines**

### **4. Are there duplicated rows?**

We are going to check that and save the result into `num_duplicated_rows`. The value of this variable is `greater than zero` if the data is duplicated and vice versa.

In [11]:
num_duplicated_rows = len(wine_df) - len(wine_df.drop_duplicates())
print('Duplicated:', num_duplicated_rows)

Duplicated: 9983


We can see the data is duplicated. Now, let's see that

In [4]:
wine_df[wine_df.duplicated()]

country                                        description  \
2408        US  This is weighty, creamy and medium to full in ...   
2409     Italy  There's a touch of toasted almond at the start...   
2410     Chile  Lightly herbal strawberry and raspberry aromas...   
2412     Italy  The Monica grape often shows a rustic or raw q...   
2413    France  Fruity and lightly herbaceous, this has fine t...   
...        ...                                                ...   
129879      US  This wine is a blend of 45% Grenache, 37% Syra...   
129880      US  Apple blossom intrigues on the nose of this li...   
129881   Spain  This Verdejo smells like citrus fruits and wil...   
129882      US  This wine is mostly Cabernet Sauvignon (78%), ...   
129913      US  This is taut and dense, and requires time and ...   

               designation  points  price           province  \
2408                   NaN      85   14.0         California   
2409    Sallier de la Tour      85   13.0  Sicily & Sardinia   
2410               Reserve      85   13.0       Maipo Valley   
2412                 Dolia      85   14.0  Sicily & Sardinia   
2413     La Fleur d'Amélie      85   15.0           Bordeaux   
...                    ...     ...    ...                ...   
129879               Metis      90   20.0         Washington   
129880                 NaN      90   20.0         California   
129881                 NaN      90   19.0     Northern Spain   
129882             Reserve      90   60.0         Washington   
129913   Artist Series #14      92   44.0         Washington   

                    region_1         region_2        taster_name  \
2408             North Coast      North Coast     Virginie Boone   
2409                 Sicilia              NaN                NaN   
2410                     NaN              NaN  Michael Schachner   
2412      Monica di Sardegna              NaN                NaN   
2413          Bordeaux Blanc              NaN         Roger Voss   
...                      ...              ...                ...   
129879  Columbia Valley (WA)  Columbia Valley   Sean P. Sullivan   
129880  Russian River Valley           Sonoma     Virginie Boone   
129881                 Rueda              NaN  Michael Schachner   
129882  Columbia Valley (WA)  Columbia Valley   Sean P. Sullivan   
129913  Columbia Valley (WA)  Columbia Valley       Paul Gregutt   

       taster_twitter_handle  \
2408                 @vboone   
2409                     NaN   
2410             @wineschach   
2412                     NaN   
2413              @vossroger   
...                      ...   
129879         @wawinereport   
129880               @vboone   
129881           @wineschach   
129882         @wawinereport   
129913           @paulgwine    

                                                    title  \
2408              Souverain 2010 Chardonnay (North Coast)   
2409    Tasca d'Almerita 2011 Sallier de la Tour Grill...   
2410    Tres Palacios 2011 Reserve Pinot Noir (Maipo V...   
2412    Cantine di Dolianova 2010 Dolia  (Monica di Sa...   
2413    Château de Sours 2011 La Fleur d'Amélie  (Bord...   
...                                                   ...   
129879  Maison Bleue 2013 Metis Red (Columbia Valley (...   
129880  Martin Ray 2015 Chardonnay (Russian River Valley)   
129881                  Martinsancho 2015 Verdejo (Rueda)   
129882   Matthews 2012 Reserve Red (Columbia Valley (WA))   
129913  Woodward Canyon 2005 Artist Series #14 Caberne...   

                           variety                winery  
2408                    Chardonnay             Souverain  
2409                        Grillo      Tasca d'Almerita  
2410                    Pinot Noir         Tres Palacios  
2412                        Monica  Cantine di Dolianova  
2413    Bordeaux-style White Blend      Château de Sours  
...                            ...                   ...  
129879                   Red Blend          Maison Bleue  
12

Having 9983 rows are duplicated. Therefore, we're going to remove these duplicated data

In [5]:
wine_df = wine_df.drop_duplicates() 
len(wine_df)

119988

### **5. What is the meaning of each column?**

The data consists of 13 fields:

- *Points*: the number of points WineEnthusiast rated the wine on a scale of 1-100 (though they say they only post reviews for wines that score >=80)
- *Title*: the title of the wine review, which often contains the vintage if you're interested in extracting that feature
- *Variety*: the type of grapes used to make the wine (ie Pinot Noir)
- *Description*: a few sentences from a sommelier describing the wine's taste, smell, look, feel, etc.
- *Country*: the country that the wine is from
- *Province*: the province or state that the wine is from
- *Region 1*: the wine growing area in a province or state (ie Napa)
- *Region 2*: sometimes there are more specific regions specified within a wine growing area (ie Rutherford inside the Napa Valley), but this value can sometimes be blank
- *Winery*: the winery that made the wine
- *Designation*: the vineyard within the winery where the grapes that made the wine are from
- *Price*: the cost for a bottle of the wine 
- *Taster Name*: name of the person who tasted and reviewed the wine
- *Taster Twitter Handle*: Twitter handle for the person who tasted ane reviewed the wine

### 6. What is the current data type of each column? Are there columns having inappropriate data types?

We check the datatype (`dtype`) of each element in a column in `wine_df` and store the results in a series named `col_dtypes`. The index of this series is the name of the columns.

In [7]:
col_dtypes = wine_df.dtypes
col_dtypes

country                   object
description               object
designation               object
points                     int64
price                    float64
province                  object
region_1                  object
region_2                  object
taster_name               object
taster_twitter_handle     object
title                     object
variety                   object
winery                    object
dtype: object

- As we can see, `points` and `price` are columns are suitable for numeric columns, the others are object type 
- There are not any inappropriate data types

### 7. With each numerical column, how are values distributed?
- What is the percentage of `missing values`?
- `Min`? `max`? Are they `abnormal`?

For now, there are 2 columns that seem to be numeric data: "points", "price". 

For each of numeric column, we will compute the percentage of missing values; min, max value and store them into a dataframe named `nume_col_profiles_df`. This dataframe contains 3 rows ("missing_ratio", "min", "max") and 2 columns ("points", "price").

In [9]:
nume_col_profiles_df = pd.DataFrame(columns = ["points", "price"], 
                                    index = ['missing_ratio', 'min', 'max'], data = np.nan)


nume_col_profiles_df.loc['missing_ratio'] = np.array(wine_df[nume_col_profiles_df.columns].isnull().sum() * 100 
                                             / len(wine_df[nume_col_profiles_df.columns]))

nume_col_profiles_df.loc['min'] = wine_df[nume_col_profiles_df.columns].min()

nume_col_profiles_df.loc['max'] = wine_df[nume_col_profiles_df.columns].max()


nume_col_profiles_df

points        price
missing_ratio     0.0     6.996533
min              80.0     4.000000
max             100.0  3300.000000

### 8. With each categorical column, how are values distributed?
- What is the percentage of `missing values`?
- How many `different values`? Show a few
- Are they `abnormal`?

`wine_df` contains 11 categorical columns: "country", "description", "designation", "province", "region_1", "region_2", "taster_name", "taster_twitter_handle", "title", "variety", "winery" .

For each of categorical column:
- compute the percentage of missing values 
- number of different values (missing values are not counted), 
- a list of different values (missing values are not counted) 

Then store them into a dataframe named `cate_col_profiles_df`. This dataframe contains 3 rows ("missing_ratio", "num_diff_vals", "diff_vals") and 11 columns ("country", "description", "designation", "province", "region_1", "region_2", "taster_name", "taster_twitter_handle", "title", "variety", "winery").

In [10]:
cate_col_profiles_df = pd.DataFrame(columns = ["country", "description", "designation", "province", "region_1", "region_2", "taster_name", "taster_twitter_handle", "title", "variety", "winery"],
                                   index = ["missing_ratio", "num_diff_vals", "diff_vals"])

cate_col_profiles_df.loc['missing_ratio'] = (wine_df[cate_col_profiles_df.columns].isnull().sum() * 100 
                                             / len(wine_df[cate_col_profiles_df.columns]))

cate_col_profiles_df.loc['num_diff_vals'] = wine_df[cate_col_profiles_df.columns].nunique()

cate_col_profiles_df.loc['diff_vals'] = wine_df[cate_col_profiles_df.columns].ffill().agg(['unique']).loc['unique']

cate_col_profiles_df

country  \
missing_ratio                                           0.049172   
num_diff_vals                                                 43   
diff_vals      [Italy, Portugal, US, Spain, France, Germany, ...   

                                                     description  \
missing_ratio                                                0.0   
num_diff_vals                                             119955   
diff_vals      [Aromas include tropical fruit, broom, brimsto...   

                                                     designation  \
missing_ratio                                          28.790379   
num_diff_vals                                              37979   
diff_vals      [Vulkà Bianco, Avidagos, Reserve Late Harvest,...   

                                                        province  \
missing_ratio                                           0.049172   
num_diff_vals                                                425   
diff_vals      [Sicily & Sardinia, Douro, Oregon, Michigan, N...   

                                                        region_1  \
missing_ratio                                           16.30163   
num_diff_vals                                               1229   
diff_vals      [Etna, Willamette Valley, Lake Michigan Shore,...   

                                                        region_2  \
missing_ratio                                          61.021936   
num_diff_vals                                                 17   
diff_vals      [nan, Willamette Valley, Napa, Sonoma, Central...   

                                                     taster_name  \
missing_ratio                                          20.766243   
num_diff_vals                                                 19   
diff_vals      [Kerin O’Keefe, Roger Voss, Paul Gregutt, Alex...   

                                           taster_twitter_handle  \
missing_ratio                                          24.540787   
num_diff_vals                                                 15   
diff_vals      [@kerinokeefe, @vossroger, @paulgwine , @wines...   

                                                           title  \
missing_ratio                                                0.0   
num_diff_vals                                             118840   
diff_vals      [Nicosia 2013 Vulkà Bianco  (Etna), Quinta dos...   

                                                         variety  \
missing_ratio                                           0.000833   
num_diff_vals                                                707   
diff_vals      [White Blend, Portuguese Red, Pinot Gris, Ries...   

                                                          winery  
missing_ratio                                                0.0  
num_diff_vals                                              16757  
diff_vals      [Nicosia, Quinta dos Avidagos, Rainstorm, St. ...

## 3. Asking meaningful questions

### Question 1

***Write question***

**Benefit:**

### Question 2

***Write question***

### **Benefit:**

### Question 3

***Write question***

### **Benefit:**

## 4. Preprocessing + analyzing data to answer each question

### Question 1:

### Preprocessing

#### Explanation

In [ ]:
# Code

### Data Analysis

#### Explanation:

In [ ]:
# Code

### Data Visualization

In [ ]:
# Code

***Comments:***

### Question 2:

### Preprocessing

#### Explanation:

In [ ]:
# Code

### Data Analysis

#### Explanation:

In [ ]:
# Code

### Data Visualization

In [ ]:
# Code

***Comments:***

### Question 3:

### Preprocessing

#### Explanation:

In [ ]:
# Code

### Data Analysis

#### Explanation:

In [ ]:
# Code

### Data Visualization

In [ ]:
# Code

***Comments***

## 5.Reflection

## 6.References